### PR TITLE
Auto user permissions creation over its deployed process

### PR DESCRIPTION
**Pull request by fmigneault**
When a process is deployed on Weaver, the user that deployed it (provided they had permission to do so), will automatically receive the necessary permissions to obtain access to its description and execution of jobs with it.

The process will be visible only by the deploying user (and admins) until it gets shared/published with other users/groups with the appropriate permissions applied in Magpie.

## Changes

**Non-breaking changes**
- Adds Magpie response hook that will create necessary permissions for user onto its deployed Weaver process.

**Breaking changes**
- n/a

**Note**: only commits after https://github.com/bird-house/birdhouse-deploy/pull/247/commits/e661ef269bdeba5d8a503bfbc5727b8c8082964e can be used to see only the diff of *this* PR omitting other dependencies.

## Related Issue / Discussion

- Resolves [DAC-206 - *user private process deployment*](https://crim-ca.atlassian.net/browse/DAC-206)

## Additional Information

- [ ] Depends on #244 
- [ ] Depends on #245
- [ ] Depends on https://github.com/crim-ca/weaver/pull/442
  (only for using the CLI below, not impacting the stack itself)

### Results / Demonstration

Using the Weaver CLI in combination of https://github.com/Ouranosinc/requests-magpie/, following command is sent and received:

```shell
weaver deploy \
  -u http://localhost:8000/ows/proxy/weaver \
  -p test-report \
  --cwl https://raw.githubusercontent.com/crim-ca/weaver/master/docs/examples/docker-python-script-report.cwl \
  --auth-handler requests_magpie.MagpieAuth \
  --auth-url http://localhost:2001 \
  --auth-username authtest \
  --auth-password authtestauthtest

[2022-06-06 16:21:35] WARNING  [weaver.utils] No settings container provided by [weaver.utils.wrapped], request options might not be applied as expected.
[2022-06-06 16:21:35] WARNING  [weaver.utils] No settings container provided by [weaver.utils.wrapped], request options might not be applied as expected.
[2022-06-06 16:21:35] WARNING  [weaver.utils] No settings container provided by [weaver.utils.wrapped], request options might not be applied as expected.
Resolved '/tmp/tmp5qbm9jep/package' to 'file:///tmp/tmp5qbm9jep/package'
{
  "description": "Process successfully deployed.",
  "processSummary": {
    "id": "test-report",
    "mutable": true,
    "keywords": [
      "application"
    ],
    "metadata": [],
    "jobControlOptions": [
      "async-execute",
      "sync-execute"
    ],
    "outputTransmission": [
      "reference",
      "value"
    ],
    "processDescriptionURL": "http://localhost:4002/processes/test-report",
    "executeEndpoint": "http://localhost:4002/processes/test-report/jobs"
  },
  "deploymentDone": true
}
```

> *note:*
    http://localhost:8000/ows/proxy/weaver is a Twitcher instance running with configured Magpie on http://localhost:2001 and Weaver defined as a service behind them.

Before execution, the following is shown in Magpie. These definitions are similar to current birdhouse stack when Weaver is enabled to be registered with minimal permissions. Deploy permission (write) is added to the test user using `match` modifier (contrary to `recursive`) such that they can POST (i.e.: deploy) on that specific endpoint, but cannot execute *all* underlying process/jobs.

image

After the execution of the CLI command, the following is obtained. This grants access to the user for execution and un-deployment of the new process, but still cannot modify/execute other processes they do not "own".

image

Logs display the valid creation:

```log
2022-06-06 16:44:19,543 INFO  [TWITCHER|magpie.adapter.magpieowssecurity:245][ThreadPoolExecutor-0_0] User <User: name=authtest id=17089> resolved with allowed 'write' access to [/ows/proxy/weaver/processes]
2022-06-06 16:44:19,544 INFO  [TWITCHER:57][ThreadPoolExecutor-0_0] Using adapter: '<class 'magpie.adapter.MagpieAdapter'>'
2022-06-06 16:44:19,546 DEBUG [TWITCHER:79][ThreadPoolExecutor-0_0] url = http://localhost:4002/processes
2022-06-06 16:44:19,712 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Using settings from local thread.
2022-06-06 16:44:19,712 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Constant alternate search: magpie.providers_hooks_path
2022-06-06 16:44:19,712 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Constant found in definitions with: MAGPIE_PROVIDERS_HOOKS_PATH
2022-06-06 16:44:19,715 DEBUG [TWITCHER|magpie.adapter:316][ThreadPoolExecutor-0_0] Hook matched response (POST /processes) [/home/francis/dev/daccs/birdhouse-deploy/birdhouse/components/weaver/config/magpie/weaver_hooks.py:allow_user_deployed_processes]
2022-06-06 16:44:19,716 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Constant alternate search: magpie.admin_group
2022-06-06 16:44:19,716 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Constant found in definitions with: MAGPIE_ADMIN_GROUP
[2022-06-06 16:44:26 -0400] [719671] [CRITICAL] WORKER TIMEOUT (pid:719703)
[2022-06-06 16:44:26 -0400] [719703] [INFO] Worker exiting (pid: 719703)
Exiting 1 (-v to see traceback)
2022-06-06 16:44:42,619 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Using settings from local thread.
2022-06-06 16:44:42,620 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Constant alternate search: magpie.anonymous_user
2022-06-06 16:44:42,620 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Constant found in definitions with: MAGPIE_ANONYMOUS_USER
2022-06-06 16:44:42,621 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Using settings from local thread.
2022-06-06 16:44:43,588 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Using settings from local thread.
2022-06-06 16:44:43,588 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Constant alternate search: magpie.anonymous_user
2022-06-06 16:44:43,589 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Constant found in definitions with: MAGPIE_ANONYMOUS_USER
2022-06-06 16:44:43,589 DEBUG [magpie.utils:163][ThreadPoolExecutor-0_0] Using settings from local thread.
2022-06-06 16:45:02,248 INFO  [birdhouse-weaver-hooks:257][ThreadPoolExecutor-0_0] Successful creation of permissions for user [authtest] to access deployed process [test-report] in Weaver.
```

### CODE REVIEW

**Comment by fmigneault** [comment]
@ahandan @francisPLT @huard 
This PR might be of interest to you, it shows a use case of `weaver` CLI used in combination with Magpie authentication on a single command, with a simply CWL reference to deploy a new process.

File: birdhouse/components/weaver/config/magpie/weaver_hooks.py
Comment (matprov) [inline birdhouse/components/weaver/config/magpie/weaver_hooks.py:56]: typo : writen

**Comment by fmigneault** [comment]
run tests

**Comment by fmigneault** [comment]
@tlvu 
I will bundle #244, #245 and this PR together once I get your approval. 
You can focus only on this change: https://github.com/bird-house/birdhouse-deploy/pull/247/files#diff-35f094ca9709a11d6e791bdd0b311be7fd4e57ce56a5c00fa42582f1f0aa10dcR172-R281
The rest you have reviewed it already.

**Review by matprov — APPROVED** [review]

**Comment by tlvu** [comment]
> @tlvu I will bundle #244, #245 and this PR together once I get your approval. You can focus only on this change: https://github.com/bird-house/birdhouse-deploy/pull/247/files#diff-35f094ca9709a11d6e791bdd0b311be7fd4e57ce56a5c00fa42582f1f0aa10dcR172-R281 The rest you have reviewed it already.

Oh thank you, I was a bit lost with all the PR with many similar code.  That change LTGM.

**Review by tlvu — APPROVED** [review]
LGTM

**Pull request merged**
